### PR TITLE
vpa-admission-controller: Improve finding the matching VPA for Pod

### PR DESCRIPTION
**Pull request by ialidzhikov**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

As outlined in https://github.com/kubernetes/autoscaler/issues/6925, right now the logic in the vpa-admission-controller to find the matching VPA for a Pod is expensive/ineffective. In a summary, no matter the given Pod is scaled by VPA or not, the vpa-admission-controller gets all VPAs in the namespace and fetches the selector for every VPA. Fetching the selector for a VPA is an expensive operation when the VPA's targetRef is a custom resource. For such cases, the `/scale` subresource is being fetched. Note that the used `selectorFetcher` does not have a cache client-side and it always does a GET request for the `/scale` subresource: https://github.com/kubernetes/autoscaler/blob/c79bdaf76210eb9ed9952da579196b87afb5665c/vertical-pod-autoscaler/pkg/target/fetcher.go#L130

This PR improves the effectiveness of the operations by using the following steps.
1. Find to the "top most well-known scalable controller" (aka parent contrller) for the Pod
1. Find all VPAs that match the parent controller
1. Fetch the selector (fetch `/scale` subresource when the targetRef is a CR) only for VPAs  which targetRef is matching the parent controller
1. Find the strongest VPA that matches the Pod 

In this way the selector would be fetched only for VPAs that match the Pod's parent controller. Previously it was fetched always for all VPAs in the namespace.

In `GetControllingVPAForPod` func it was always fetching the parent controller for a Pod, hence we are not introducing a change by fetching always the Pod's parent controller at the beginning of the `GetMatchingVPA` func.

#### Which issue(s) this PR fixes:

N/A
but a result of the following issues https://github.com/kubernetes/autoscaler/issues/6925, https://github.com/kubernetes/autoscaler/issues/6884

#### Does this PR introduce a user-facing change?

```release-note
vpa-admission-controller: The logic for finding the matching VPA for Pod is now more effective.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by plkokanov — COMMENTED** [review]
Thanks for opening the PR! Just one small comment on my side.

File: vertical-pod-autoscaler/pkg/admission-controller/resource/vpa/matcher.go [hunk 1]
```diff
@@ -50,30 +50,46 @@ func NewMatcher(vpaLister vpa_lister.VerticalPodAutoscalerLister,
...
 		return nil
 	}
-	onConfigs := make([]*vpa_api_util.VpaWithSelector, 0)
+
+	var controllingVpa *vpa_types.VerticalPodAutoscaler
 	for _, vpaConfig := range configs {
 		if vpa_api_util.GetUpdateMode(vpaConfig) == vpa_types.UpdateModeOff {
 			continue
 		}
+		if vpaConfig.Spec.TargetRef == nil {
```
[the remark this answers is not in the document]
Comment (ialidzhikov) [reply vertical-pod-autoscaler/pkg/admission-controller/resource/vpa/matcher.go:73]: I added an info level logging with a lower verbosity (5), see https://github.com/kubernetes/autoscaler/pull/7010/commits/20dc06b96c3e1e8b7bf7bba6fcc7c33931b35590.

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go [hunk 2]
```diff
@@ -109,8 +109,8 @@ func PodLabelsMatchVPA(podNamespace string, labels labels.Set, vpaNamespace stri
 	return vpaSelector.Matches(labels)
 }
 
-// stronger returns true iff a is before b in the order to control a Pod (that matches both VPAs).
-func stronger(a, b *vpa_types.VerticalPodAutoscaler) bool {
+// Stronger returns true iff a is before b in the order to control a Pod (that matches both VPAs).
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/utils/vpa/api.go:112]:
```suggestion
// Stronger returns true if a is before b in the order to control a Pod (that matches both VPAs).
```

A small typo  in the original comment

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go [hunk 3]
```diff
@@ -162,21 +143,46 @@ func GetControllingVPAForPod(pod *core.Pod, vpas []*VpaWithSelector, ctrlFetcher
...
+		if PodMatchesVPA(pod, vpaWithSelector) && Stronger(vpaWithSelector.Vpa, controllingVpa) {
 			controlling = vpaWithSelector
 			controllingVpa = controlling.Vpa
 		}
 	}
 	return controlling
 }
 
+// FindParentControllerForPod returns the parent controller (topmost well-known or scalable controller) for the given Pod.
+func FindParentControllerForPod(pod *core.Pod, ctrlFetcher controllerfetcher.ControllerFetcher) (*controllerfetcher.ControllerKeyWithAPIVersion, error) {
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/utils/vpa/api.go:163]: Can a test be written for this function?

**Comment by ialidzhikov** [comment]
I am sorry for the delay. I got stuck with other high priority tasks. I will get back to the PR review and address it next week.

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go (hunk 2 above)
Comment (ialidzhikov) [reply vertical-pod-autoscaler/pkg/utils/vpa/api.go:112]: `iff` is actually used in many places in the source code. I am not sure if it is a typo or refers to `if and only if (as a written abbreviation).`

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go (hunk 3 above)
Comment (ialidzhikov) [reply vertical-pod-autoscaler/pkg/utils/vpa/api.go]: I added unit test in https://github.com/kubernetes/autoscaler/pull/7010/commits/dcbd7164678527fb9ed867c20c4a74215d026c38.

**Comment by ialidzhikov** [comment]
@adrianmoisey @plkokanov @voelzmo I rebased the PR and address review comments! PTAL.

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go (hunk 2 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/utils/vpa/api.go:112]: That's fair, I think you're right

**Review by plkokanov — APPROVED** [review]
lgtm on my side

**Comment by adrianmoisey** [comment]
I've had a look and I'm also generally happy with it
I'd like to still give it a test run locally

**Comment by adrianmoisey** [comment]
I gave this a test, seems to work.
I'm unsure how to compare the performance of this change to what it was before hand though.
/lgtm

**Pull request merged**